designdocs/recovery_system.md

```diff
@@ -0,0 +1,505 @@
+    B->>SDK: Return addProviderPayload, claimHandlePayload, graph key setup payload, recoveryTokenPayload
+    Note over App,SDK: Signature request may need to be mediated by App
+    SDK->>EW: Request signatures for payloads
+    EW->>SDK: Return signatures
+    SDK->>B: Submit signed payloads
+    B->>G: *NEW* Create MSA with recovery email
+    G->>B: Return generated graph key and recovery code
+    G->>F: Submit batch transaction
+    B->>+G: (Polling) Get MSA
+    G->>F: Check for MSA
+    G->>-B: MSA Confirmation
+    B->>B: Store MSA, Graph Key, Recovery Code
+    B->>G: Get full handle
+    G->>F: handleForMSA
+    F->>G: Return handle
+    G->>B: Return handle
+    B->>App: Return handle
+    App->>U: Show confirmed handle
+    G->>U: (Async) Send Frequency intro/recovery email including recovery code
+```
+
+#### Sequence diagram – Account creation and recovery key setup
+
+1. User Account Creation: The user signs up with a wallet/app (the provider) and provides their email address as part of registration. The provider creates a new MSA on the Frequency blockchain for the user. During this on-chain account creation, the user’s initial control key (public key) is linked to the MSA (this may be a key managed by the provider).
+
+2. Recovery Key Generation: The blockchain generates a random secret (e.g. a 128-bit entropy value). For example, it may call a secure random generator to get 16 bytes and then encode as a 12-word BIP-39 phrase for usability, or a UUIDv4 code. This secret is combined with the user’s email (e.g. by hashing as described in the Cryptographic Design) to produce the recovery hash record. The blockchain implements the hash option, e.g. it computes H = SHA-256(email || recovery_code || global salt).
+
+3. On-Chain Storage: The blockchain stores the recovery hash record (H) as part of the MSA creation extrinsic or a separate call right after. The blockchain will store H in association with that MSA ID. If this is part of the creation, the “Create MSA” call would include the hashed recovery key as an attribute. The chain may also stake the provider’s identity to ensure only authorized providers can do this (since only a registered provider should write recovery keys, protecting users from malicious data injection).
+
+4. User Receives Recovery Code: Once the MSA is successfully created and the recovery hash stored, the provider delivers the actual recovery code to the user. Typically, the wallet UI would display the 12-word phrase or UUID code and prompt the user to save it. For security, the provider does not send this over email (to avoid having the secret in the inbox). The user is instructed: “Save this recovery phrase offline. It will allow you to recover your account if needed.” This two-factor approach is still secure if the recovery code is sent via email, as long as the email verification step is robust. It is however, more secure to avoid sending the recovery code via email entirely.
```

Comment:
>  hash stored, the provider delivers the
Slight modification, I think you mean recovery providers?
 hash stored, the _recovery_ provider delivers the

Reply:
> Actually, are Recovery Providers different than Regular providers? Can providers be both a Regular Provider and Recovery Provider?

I would say currently there is nothing that would prevent a provider from being just one or both. Don't think there needs to be an exclusivity here?

Reply:
Thanks! Like any design decision, there are tradeoffs. Having a provider take on both roles can increase centralization and the level of trust placed in that provider. That said, since Recovery Providers are approved through governance, theres already an additional layer of vetting in place which helps mitigate risk.